Adds the simple-linked-list exercise (#311)

joelwallis opened the pull request:
I needed to implement this as a new branch since the work in progress of #311 was in the contributor's fork. This is my first exercise contribution, so please consider that I don't exactly know what I am doing 😂

Suggestions for improvements are more than welcome! 🎉

joelwallis commented:
![](https://media.giphy.com/media/LqW9dLVjQm3cs/giphy.gif)

joelwallis commented:
Forgot to add the exercise to `config.json`. Now tests are passing.

Also, I'm not sure which difficulty I should attribute to it, but I think `1` is not a fair one. A beginner would struggle to implement this one. Also, any suggestion for topics?

matthewmorgan commented on exercises/simple-linked-list/example.js:
```diff
@@ -0,0 +1,72 @@
+function ListElement (value, next) {
+  if (!(this instanceof ListElement))
```
Let's use curly braces here, even though they're not strictly needed.  It's the style we've been using.
Same elsewhere-- I won't call out each spot.

joelwallis replied:
Roger that!

Also, where exactly did we set our coding standards? I remember we had a discussion about it but I can't remember if we decided to adopt a specific one.

matthewmorgan commented on exercises/simple-linked-list/example.js:
```diff
@@ -0,0 +1,72 @@
+
+  var newEl = (value instanceof ListElement)
+    ? value
+    : new ListElement(value)
```
Semicolon at the end of this line.

joelwallis replied:
Should I use semicolons to end statements. If that's the case, I'll add them across the file. I think I've coded it mostly following the JS Standard Style (no problem on changing this, just want to make sure).

matthewmorgan commented on exercises/simple-linked-list/example.js:
```diff
@@ -0,0 +1,72 @@
+  while (lastEl.next) lastEl = lastEl.next;
+
+  lastEl.next = newEl;
+}
```
Semicolon after here too.  Again, I won't call it out everywhere, but our current style is to use them by default.

matthewmorgan commented on exercises/simple-linked-list/simple-linked-list.spec.js:
```diff
@@ -0,0 +1,129 @@
+    expect(LinkedList).toBeDefined();
+  });
+
+  it('exports a ListElement constructor', function () {
```
don't forget to `xit` all the tests except the first one :)

matthewmorgan commented on exercises/simple-linked-list/example.js:
```diff
@@ -0,0 +1,72 @@
+function ListElement (value, next) {
```
So, this is a nitpick, but the canonical README says:

> Write a simple linked list implementation that uses Elements and a List.

I think with that in mind, the classes should be changed to be called `Element` and `List`.  Since these aren't reserved words in JavaScript, we should be OK.

joelwallis replied:
`Element` and `List` would be much better! Change is on its way

joelwallis commented:
Well, that's weird. I've just pushed the requested changes but the GitHub Review doesn't reflect it. Probably because I used `--amend` while committing.

Anyway, it's available here: https://github.com/exercism/javascript/commit/59419edd54e59ce1cd03efcfb538c751962d5d20

joelwallis commented:
Hey @matthewmorgan!

I added tests for `List#reverse`, `List.toArray` and `List#toArray` methods, alongside an implementation example. Also, I changed the difficulty from 1 to 8. Let me know if there's anything else missing. 😊

matthewmorgan commented on exercises/simple-linked-list/example.js:
```diff
@@ -0,0 +1,119 @@
+  }
+
+  lastEl.next = newEl;
+}
```
I hate to even mention this on your beautiful pull request, but all these function expressions should have terminating semicolons...this is why we need a linter!  Would you mind adding them in the meantime?

joelwallis replied:
Hey @matthewmorgan, no need feel bad about it. Regarding the function expression, you meant all function declarations? Named and unnamed ones? E.g.:

```js
// named function
function foo () {}
// anonymous function
var foo = function () {}
```

So, both of them should end with semicolons, right?

```js
// named function
function foo () {};
// anonymous function
var foo = function () {};
```

joelwallis replied:
Well, I just pushed this change, assuming that both named and anonymous should receive semicolons.

matthewmorgan commented:
This looks terrific.  Nice work sir!

If you wouldn't mind adding some more semicolons in a couple places...this is otherwise ready to merge!

joelwallis commented:
@matthewmorgan done! I'll wait for the CI to complete the tests, then I'll merge this PR. Thank you very much for your review!

matthewmorgan commented:
@joelwallis Nice job!  Thanks for doing this.

rchavarria commented:
I'm too late to reply but here are my two cents: I agree on the difficulty level and the topic. The bad news is that the topic doesn't exist in the [topic list](https://github.com/exercism/problem-specifications/blob/master/TOPICS.txt) (is it official?).

Anyway, I think the better topic to describe the exercise is `Data structures`, so it would be a matter of propose it to be added to the list. If we want to add more topics, I would add: `Arrays`, `Lists`.

matthewmorgan commented:
@rchavarria thanks for catching that!  @joelwallis would you mind creating a follow-up issue for that?

joelwallis commented:
@matthewmorgan definitely not! @rchavarria thanks for catching! I'll PR it right now
